### DISCUSSION TITLE
anyway to only allow predefined list of field names in Designer?

### DISCUSSION THREAD

**Discussion by mattfeury** [pdfme/pdfme#419] [Q&A]
hello! lovely lib you all have made.

for my use case, i'd like to only allow a user in the "Designer" to name fields from a predefined list of name options (the use case being i want to map these fields to certain generated values on the backend). is there any precedent for doing this?

one solution seems like it could be to create a new schema and just do it all myself, but it seems like the actual "naming" piece of things is handled outside of the schema itself? i'm not even sure it's the right approach.

the hackier approach would be just to fork the lib and modify the field name "input" to be a select instead of a basic text field.

i figured i'd ask here to see if there was something i'm missing before going down any particular path.

thanks!

**Comment by mattfeury** [comment]
after poking around a bit more, it seems like my best bet here may be to do a custom plugin that just disregards the name alltogether and uses the Plugin's element value to trigger a select box. 

will leave this open in case other folks want to chime in with another take

**Comment by hand-dot** [comment]
Hey @mattfeury 

Currently, the field names are automatically determined by the following code:

[https://github.com/pdfme/pdfme/blob/25077a1a05a0c022b085ac7e731612ed51e411dc/packages/ui/src/components/Designer/index.tsx#L164](https://github.com/pdfme/pdfme/blob/25077a1a05a0c022b085ac7e731612ed51e411dc/packages/ui/src/components/Designer/index.tsx#L164)

I also didn't think the automatically generated field names provided a good user experience. Using plugin/schema settings to decide on the field names is a great idea. One thing to be mindful of is that the field names need to be unique, so it might be necessary to come up with a scheme, such as appending numbers to the end.

On a slightly different note...

Although implementation has not started yet, there is a plan to allow adding schemas via drag and drop from the left sidebar. [https://github.com/pdfme/pdfme/issues/400](https://github.com/pdfme/pdfme/issues/400) For this implementation, I was thinking of adding an icon property to the schema definition. It seems possible to also include the addition of a fieldName property with this update.

**Reply by mattfeury** [reply]
hey @hand-dot thanks for the response. i implemented my solution and it's working fairly well. one q though: is there anything actually requiring the field names be unique? i see when i put a duplicate name on a single page that it shows a message, which is nice (though it still lets me save). however, i have a multi page pdf and it appears that each page has its own field list. however, i can name the same field the same thing on page1 and page2 and no warning is given. this also is problematic because when i pass in inputs, it can only use one value.

**Reply by mattfeury** [reply]
example: i create a new field on page1 and it defaults to `field1`. i create a new field on page2, and it also defaults to `field1`. no warning is given. if i then attempt to generate PDF, i can only pass in a single input value `{ field1: "value" }`. is this a bug?

**Reply by jkugsiya** [reply]
> hey @hand-dot thanks for the response. i implemented my solution and it's working fairly well. one q though: is there anything actually requiring the field names be unique? i see when i put a duplicate name on a single page that it shows a message, which is nice (though it still lets me save). however, i have a multi page pdf and it appears that each page has its own field list. however, i can name the same field the same thing on page1 and page2 and no warning is given. this also is problematic because when i pass in inputs, it can only use one value.

Hey @mattfeury and @hand-dot , I also need something like this or with a little more tweak, for ex: I have some default templates, the names which are loaded from the default templates should not be editable (the input should be disabled in this case and also deletion of these fields should also be restricted). Also your solution would also help me that we can only choose from the given list of names, can you please help me here.
